### ISSUE TITLE
Unnecessary temporaries generated for some tuple assignments

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#50784]
**Version Used**: 
https://github.com/dotnet/roslyn/commit/37a652243169f3a7fdf78fc380f53d0b87eecf23

**Steps to Reproduce**:
https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEDMACFAmDBhAWACgBvYjCjABwQEsA3CeDWgOxgwH0IAaL4ANzlKwiuiwAWDAFkAjAAo2HXi3YZgASlEYyRSvozzufTpowBeQys1C9lAL7bt4lFOnZ5WuxV0HKSjAgLLmtg01s/CiMVUw1g+WsNCIdieyA
```C#
public class C
{
    private int _a, _b;
    
    public void M1(int a, int b)
    {
        (_a, _b) = (a, b);
    }
    
    public void M2()
    {
        int a = _a, b = _b;
        (_a, _b) = (a, b);
    }
}
```

**Expected Behavior**:
The compiler produces IL equivalent to:
```C#
public class C
{
    private int _a, _b;

    public void M1(int a, int b)
    {
        _a = a;
        _b = b;
    }

    public void M2(int a, int b)
    {
        int a = _a, b = _b;
        _a = a;
        _b = b;
    }
}
```

**Actual Behavior**:
The compiler produces IL equivalent to:
```C#
public class C
{
    private int _a, _b;

    public void M1(int a, int b)
    {
        _a = a;
        _b = b;
    }

    public void M2()
    {
        int a = _a;
        int b = _b;
        int a2 = a;
        int b2 = b;
        _a = a2;
        _b = b2;
    }
}
```

**Comment by Youssef1313** [comment]
Related to (but not duplicate of) https://github.com/dotnet/roslyn/issues/48052

**Comment by jaredpar** [comment]
The reason for the extra temporaries is that the compiler is attempting to prevent side effects from unintentionally breaking the intended semantics here. Couple of example cases where we can't do this type optimization:

```cs
// Once complete _a needs to have the original value of a, not b hence 
// we can't just evaluate the RHS and then assign the values over
(_a, _b) = (a, a = b);

// Sneaky case where even just all locals requires temporaries to evaluate
// correctly.
(a, b) = (b, a); 
```

That being said there are a number of cases where we could do this and those cases likely have some good overlap with common patterns here. We should see if we can get those common cases handled here (assuming the complexity of the check isn't too high).

**Comment by jcouv** [comment]
@RikkiGibson has a PR out to improve this for cases were it is safe (ie. compatible with spec'ed evaluation order and therefore unobservable)

**Comment by RikkiGibson** [comment]
Yeah, I think we mainly need to figure out the right heuristics. My PR as written won't handle assignments to fields at all. Planning to follow up offline about it.
